### PR TITLE
[RFC] Shell: support bell and buffer incomplete UTF-8 sequences

### LINKED ISSUE #4338: Zero width control characters insert spaces with shell ex commands
- Neovim version: 9a3b7fa90
- Operating system: Gentoo 2.2
- Terminal emulator: iTerm 2.1.1 (connected to remote system)

Discovered while researching https://github.com/neovim/neovim/issues/4142 and https://github.com/neovim/neovim/issues/2958

Running a shell command that echoes multiple bell symbols followed by text shifts text to the right.
### Actual behaviour

Text preceded by zero width control bell character is shifted to the right.
### Expected behaviour

Text should start in the first column.
### Steps to reproduce

```
nvim -u NONE
:!echo -en '\x07\x07\x07\x07text'
// prints
    text
// instead of
text
```

vim 7.4.769 with same configuration exhibits expected behavior

### PR DESCRIPTION
**Pull request by bfredl**
actually fixes #4338. I will take a peek at gvim source code if there is other low hanging fruit. But more complex stuff (like colors) should be punted until `:terminal` is integrated.

I also merged bang_filter_spec into output_spec as in fact only tests bang. Filter is tested elsewhere.

### CODE REVIEW

**Comment by bfredl** [comment]
Actually our `execute('!ls')` behavior is consistent with gvim which is inconsistent with terminal vim.
gvim does one more thing: it makes sure to buffer an incomplete multibyte character, we should do that as well.

**Comment by janlazo** [comment]
@bfredl Can you enable the tests from bang_filter_spec.lua on Windows? I did it in https://github.com/janlazo/neovim/commit/6e077b20ac27aa1815d7c7cf643a22e0d9df2db0#diff-8cc34a316e32ee95eb86a21b308e7082

**Comment by bfredl** [comment]
Thanks, I cherry picked it.

**Comment by bfredl** [comment]
Failures looks unrelated (nodejs...), marking RFC.

File: test/functional/ui/output_spec.lua [hunk 1]
```diff
@@ -131,4 +139,91 @@ describe("shell command :!", function()
     ]])
     feed([[<CR>]])
   end)
+
+  describe('', function()
```
Comment (justinmk) [inline test/functional/ui/output_spec.lua]: describe("shell command :!", function()

Comment (justinmk) [reply test/functional/ui/output_spec.lua]: My mistake, it looked like the same indent level as line 80.  I would normally put it in its own scope but doesn't matter I guess.

**Pull request merged**